### ISSUE TITLE
UI anims play at the wrong speed when FPS is not set to 60

### ISSUE THREAD

**Issue by Sushiman3000** [ikemen-engine/Ikemen-GO#2131]
### Describe the bug

Whenever you set your FPS (Framerate setting) in config.json above or below 60, UI animations will play at the wrong speed.

Setting the Framerate to match your monitor’s refresh rate makes the game run smoothly on high Hz monitors but causes UI animations to play at twice the speed. This issue does not affect characters, explods, stages, or other gameplay elements, only UI animations.

Here’s an example using a character animation as a UI animation (char portrait/face on the select screen):

![Honeycam 2024-11-19 16-12-04](https://github.com/user-attachments/assets/a01a53a0-9948-4445-9f76-e42d36d477b0)

### To Reproduce

1. Set the "Framerate" setting in save/config.json to 120 or 30.
2. Background, power bars, and other animated UI elements will play at the wrong speed.

### Expected behavior

Keep UI animations locked at 60 FPS, independent of the "Framerate" setting.

### Engine Version (or source code date)

nightly (9/11/24)

### Operating system

Windows

**Comment by K4thos** [comment]
edit: need to investigate it further

**Comment by K4thos** [comment]
works as expected, that setting affects the rate at which consecutive frames are rendered. Including menu frames. Not sure if we should even externalize it.

**Comment by K4thos** [comment]
Ok, I see where the problem lies. When you set the Framerate to 120, the entire game runs twice as fast internally, but the in-game speed is halved due to the GameFramerate setting. The game speed is calculated as GameFramerate / Framerate, which in this case equals 0.5. We could potentially use this behavior for screenpack rendering as well.

However, I’m not sure if the Framerate setting should be externalized in the first place. Does running the game with Framerate set to 120 and halving the game speed via GameFramerate set to 60 really make a noticeable difference on a high-refresh-rate monitor compared to Framerate = 60 and GameFramerate = 60?

**Comment by Sushiman3000** [comment]
It really makes a difference. You can compare it to setting the Framerate to 30 while keeping GameFramerate at 60 on a 60Hz monitor. That's how Framerate at 60 looks on a 120Hz+ monitor, though not as bad, it’s still quite noticeable. I’m referring to in-game performance, while still running at normal speed, there’s a lot of micro-stuttering.

This isn’t an Ikemen only issue though, but a general fps/hz thing issue on all games. 60fps looks smooth on a 60Hz monitor but looks awful on a 120Hz+ monitor. That’s why I think this change would be useful to people with 120Hz or higher monitors, so they can have smooth in-game performance while keeping normal speed UI elements in custom screenpacks with lots of animations.

I can try to provide some examples if that helps.

**Comment by potsmugen** [comment]
For some things it doesn't make a difference, but with characters for instance Ikemen will deliberately interpolate their movement to make it appear smoother. Lifebars too. The easiest way to see it is setting debug speed to 25%. Chars will still move at 60fps, just slower.

I don't particularly care to set frame rate above 60, but I'm sure many people would get a kick out of running Ikemen like that. Actually I can picture screenpacks with scrolling elements and such also benefitting from it, but that'd require porting the position interpolation to their code.

**Comment by potsmugen** [comment]
This seems to happen at least in part because BGDef.action() updates animations every frame, unlike stage.action() which uses the engine's tickFrame/tickNextFrame logic to check if the frame is a logic frame (where things should update) or an inbetween.

**Comment by potsmugen** [comment]
This still happens with many things. The screenpack BG's act normal but for instance the char animations update at the video rate instead of logic rate. The screenpack logic itself also updates at the wrong speed, which is easy to check with demo mode. If you make FrameRate 10 times faster demo mode will start (and end) 10 times sooner.

**Comment by UrsoManso** [comment]
Is there any chance this gets fixed before 1.0? I used to play with higher internal framerate but had to go back to 60 because of this issue.

**Comment by K4thos** [comment]
No. It would likely require a larger refactor of Lua than what was done recently, and the last one was already a mess to stabilize.

**Comment by potsmugen** [comment]
I have a couple ideas about how to patch it, but at the same time am afraid that patch would prevent us from ever fixing it properly.

**Comment by rakieldev** [comment]
A little update on this: I couldn't reproduce it on the select screen. The animations don't seem to be changing speed for me anymore, so maybe someone else could test it to confirm. But somehow it does affect the pause menu, depending on the fps setting, you can see the box cursor animation and tween getting sped up or slowed down.
